django/thunderstore/frontend/api/experimental/tests/test_package_details.py

```diff
@@ -0,0 +1,192 @@
+    dependency.package.latest.dependants.set(
+        [PackageVersionFactory(), PackageVersionFactory()]
+    )
+
+    data = __query_api(api_client, dependency)
+
+    assert data["dependant_count"] == 2
+
+
+@pytest.mark.django_db
+def test_downloads_count_is_annotated_correctly(api_client: APIClient) -> None:
+    listing = PackageListingFactory(package_version_kwargs={"downloads": 0})
+    PackageVersionFactory(package=listing.package, downloads=1, version_number="1.0.1")
+    PackageVersionFactory(package=listing.package, downloads=2, version_number="1.0.2")
+
+    data = __query_api(api_client, listing)
+
+    assert data["download_count"] == 3
+
+
+@pytest.mark.django_db
+def test_rating_is_annotated_correctly_when_no_ratings(api_client: APIClient) -> None:
+    listing = PackageListingFactory()
+
+    data = __query_api(api_client, listing)
+
+    assert data["rating_score"] == 0
+
+
+@pytest.mark.django_db
+def test_rating_is_annotated_correctly_when_ratings(api_client: APIClient) -> None:
+    listing = PackageListingFactory()
+    PackageRatingFactory(package=listing.package)
+    PackageRatingFactory(package=listing.package)
+
+    data = __query_api(api_client, listing)
+
+    assert data["rating_score"] == 2
+
+
+@pytest.mark.django_db
+def test_dependencies_for_correct_version_are_returned(api_client: APIClient) -> None:
+    listing = PackageListingFactory()
+    old = listing.package.latest
+    newer = PackageVersionFactory(package=old.package, version_number="2.0.0")
+    dependency1 = PackageListingFactory()
+    dependency2 = PackageListingFactory()
+    old.dependencies.set([dependency1.package.latest])
+    newer.dependencies.set([dependency2.package.latest])
+
+    data = __query_api(api_client, listing)
+
+    assert len(data["dependencies"]) == 1
+    assert data["dependencies"][0]["package_name"] == dependency2.package.name
+
+
+@pytest.mark.django_db
+def test_dependencies_refer_to_preferred_community(api_client: APIClient) -> None:
+    alternative_dependency = PackageListingFactory()
+    preferred_dependency = PackageListingFactory(package=alternative_dependency.package)
+    dependant = PackageListingFactory(community=preferred_dependency.community)
+    dependant.package.latest.dependencies.set([alternative_dependency.package.latest])
+
+    data = __query_api(api_client, dependant)
+
+    assert len(data["dependencies"]) == 1
+    assert data["dependencies"][0]["package_name"] == preferred_dependency.package.name
+    assert (
+        data["dependencies"][0]["community_identifier"]
+        == preferred_dependency.community.identifier
+    )
+
+
+@pytest.mark.django_db
+def test_dependencies_fallback_to_any_community(api_client: APIClient) -> None:
+    dependency = PackageListingFactory()
+    dependant = PackageListingFactory()
+    dependant.package.latest.dependencies.set([dependency.package.latest])
+
+    data = __query_api(api_client, dependant)
+
+    assert len(data["dependencies"]) == 1
```

Comment:
I feel like this test should assert the two package listings are in different communities somehow, otherwise we don't really know if it's falling back to the non-preferred community or if they just happen to be in the same community

Reply:
The way the dependant/dependency are initiated with parameterless factories guarantees they're in different communities, but that's not visible from reading the test case, and someone can always change how the factory works, so I'll add a check just in case.